src/app/hooks/useGlean.ts

```diff
@@ -4,46 +4,116 @@
 
 "use client";
 
-import { useEffect } from "react";
+import { useEffect, useState } from "react";
 import Glean from "@mozilla/glean/web";
-import * as pageEvents from "../../telemetry/generated/page";
+// @ts-expect-error We do not have type declarations for metrics.yaml, yet.
+import gleanMetrics from "../../telemetry/metrics.yaml";
+import {
+  SnakeToCamelCase,
+  convertSnakeToCamelCase,
+} from "../functions/universal/convertSnakeToCamelCase";
+
+const APP_ENV = process.env.NEXT_PUBLIC_APP_ENV || "";
+
+export type GleanExtraKeys = Record<string, string | boolean | number>;
+
+type GleanCategoryKey = keyof typeof gleanMetrics;
+
+type GleanEvent = {
+  [key: string]: {
+    type: string;
+    description: string;
+    bugs: string;
+    data_reviews: string;
+    data_sensitivity: string;
+    notification_emails: string;
+    expires: string;
+    extra_keys: GleanExtraKeys;
+    record: (params: GleanExtraKeys) => void;
+  };
+};
+
+type GleanCategories = {
+  [key: GleanCategoryKey]: GleanEvent;
+};
+
+type GleanMetrics = {
+  $schema: string;
+} & GleanCategories;
+
+// We do not need $schema here — ignore it.
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const { $schema, ...metrics }: GleanMetrics = gleanMetrics;
+
+type GleanMetricModules =
+  | {
+      [K in keyof GleanCategories as SnakeToCamelCase<string>]: GleanCategories[K];
+    }
+  | null;
 
 // Custom hook that initializes Glean and returns the Glean objects required
 // to record data.
-export const useGlean = (channel: string, appEnv: string) => {
-  // Initialize Glean only on the first render
-  // of our custom hook.
+export const useGlean = () => {
+  const [gleanEvents, setGleanEvents] = useState<GleanMetricModules>(null);
+
+  const loadGleanModules = () => {
+    const gleanModuleNames = Object.keys(metrics).map((metricCategory) =>
+      convertSnakeToCamelCase(metricCategory),
+    );
+    const modulePromises = gleanModuleNames.map(
+      (moduleName) => import(`../../telemetry/generated/${moduleName}`),
+    );
+
+    Promise.all(modulePromises)
+      .then((importedModules) => {
+        const exportModules = gleanModuleNames.reduce(
+          (modules, moduleName, moduleIndex) => ({
+            ...modules,
+            [moduleName]: importedModules[moduleIndex],
+          }),
+          {},
+        );
+        setGleanEvents(exportModules);
+      })
+      .catch((error) => {
+        console.log("Failed to load Glean modules:", error);
```

Comment:
I assume we might want to report those to Sentry.

Reply:
Yes I agree, `captureException(error)` would be appropriate here.